### PR TITLE
US040 Adding a Login System - UI

### PR DESCRIPTION
**Pull request by CharlesM988**
Adding a login page to the application. 

When a user logs in the application updates all 'login' links to 'logout' links.

Created:

- auth.ts
- authController
- config.ts
- login.html
- LoginTest.js

Updated:
- jobController.ts
- app.ts
- styles.css

Screenshot 2024-01-09 at 10 28 08
Screenshot 2024-01-09 at 10 28 19
Screenshot 2024-01-09 at 10 28 30
Screenshot 2024-01-09 at 10 27 58

### CODE REVIEW

**Review by caitlinbulloch — APPROVED** [review]
everything looks good to me 👍 just a few conflicts

File: Test/ui/LoginTest.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+const chai = require("chai");
+
+const webdriver = require("selenium-webdriver");
+
+describe("LoginTest", () => {
+  it("should click login, try to log in and generate a fail", async () => {
+    let driver = await new Builder().forBrowser("chrome").build();
+
+    try {
+      await driver.get("http://localhost:3000/");
```
Comment (eoinmcbennett) [inline Test/ui/LoginTest.js:11]: This wont work if ran in AWS

File: app.ts [hunk 2]
```diff
@@ -27,20 +29,23 @@ app.use("/public", express.static(path.join(__dirname, "public")));
 
 app.use(express.json());
 
-app.use(express.urlencoded({ extended: false }));
+app.use(bodyParser.urlencoded({ extended: true }));
+
+app.use(session({
+    secret: process.env.APP_SECRET_KEY || config.secretKey, cookie: { maxAge: 60000,
```
Comment (eoinmcbennett) [inline app.ts]: reference the env variable inside the config, it shouldn't be ignored if your using env variables

File: app.ts [hunk 3]
```diff
@@ -27,20 +29,23 @@ app.use("/public", express.static(path.join(__dirname, "public")));
...
+        token: string;
     }
 }
 
 app.listen(3000, () => {
     console.log("Server running on port 3000");
 });
 
-// Express Routes
-app.get("/", async (req: Request, res: Response) => {
```
Comment (eoinmcbennett) [inline app.ts:42]: Was this intended?

File: config.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import { randomBytes } from "crypto";
+
+const generateRandomKey = () => {
+    const randomBytesResult = randomBytes(32);
+    return randomBytesResult.toString("hex");
+};
+
+const config = {
+    secretKey: process.env.APP_SECRET_KEY || generateRandomKey()
```
Comment (eoinmcbennett) [inline config.ts]: You can just reference the config in your app.ts since you use env variables in both places

File: controller/authController.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+            req.session.isLoggedIn = true;
+            res.redirect("/job-roles");
+        } catch (e) {
+            console.error(e);
+            res.locals.errormessage = e.message;
+            res.render("login", req.body);
+        }
+    });
+
+    app.get("/logout", async (req: Request, res: Response) => {
```
Comment (eoinmcbennett) [inline controller/authController.ts]: This should be post

File: controller/jobController.ts [hunk 6]
```diff
@@ -1,37 +1,61 @@
...
 
 const jobService = require("../service/jobService");
+const authService = require("../service/authService");
+const session = require("express-session");
+
+interface CustomSessionData extends SessionData {
+    isLoggedIn: boolean;
+}
+
+declare module "express-session" {
```
Comment (eoinmcbennett) [inline controller/jobController.ts]: Is there multiple of these?

File: controller/jobController.ts [hunk 7]
```diff
@@ -1,37 +1,61 @@
...
+    interface SessionData {
+        isLoggedIn: boolean;
+    }
+}
 
 module.exports = function(app: Application){
+    app.get("/", async (req: Request, res: Response) => {
+        res.render("index", { isLoggedIn: req.session.isLoggedIn, title: "Home" });
+    });
+
```
Comment (eoinmcbennett) [inline controller/jobController.ts]: Does this really belong here?

File: controller/jobController.ts [hunk 8]
```diff
@@ -1,37 +1,61 @@
...
     app.post("/index", async (req: Request, res: Response) => {
         res.redirect("job-roles");
     });
```
Comment (eoinmcbennett) [inline controller/jobController.ts]: Why is this needed?

File: controller/jobController.ts [hunk 9]
```diff
@@ -1,37 +1,61 @@
...
 
+    app.get("/login", async (req: Request, res: Response) => {
+        res.render("login");
+    });
+
```
Comment (eoinmcbennett) [inline controller/jobController.ts]: Is this not already in your auth controller?

File: controller/jobController.ts [hunk 10]
```diff
@@ -1,37 +1,61 @@
...
     app.get("/job-roles", async (req: Request, res: Response) => {
         let data: Job[] | undefined;
+        let isLoggedIn: boolean = req.session.isLoggedIn;
 
         try {
             data = await jobService.getJobs();
+            isLoggedIn = req.session.isLoggedIn;
```
Comment (eoinmcbennett) [inline controller/jobController.ts]: No point assigning this again

File: controller/jobController.ts [hunk 11]
```diff
@@ -1,37 +1,61 @@
...
         res.redirect("job-roles");
     });
 
+    app.get("/login", async (req: Request, res: Response) => {
+        res.render("login");
+    });
+
     app.get("/job-roles", async (req: Request, res: Response) => {
         let data: Job[] | undefined;
+        let isLoggedIn: boolean = req.session.isLoggedIn;
```
Comment (eoinmcbennett) [inline controller/jobController.ts]: const

File: controller/jobController.ts [hunk 12]
```diff
@@ -1,37 +1,61 @@
...
 
         try {
             data = await jobService.getJobs();
+            isLoggedIn = req.session.isLoggedIn;
         } catch (e) {
             console.error(e);
         }
-        
-        res.render("list-jobs", { job: data } );
+        res.render("list-jobs", { job: data, isLoggedIn: req.session.isLoggedIn } );
```
Comment (eoinmcbennett) [inline controller/jobController.ts]: Use your isLoggedIn variable for this

File: controller/jobController.ts [hunk 13]
```diff
@@ -1,37 +1,61 @@
...
     });
 
     app.get("/list-jobs/:id", async (req: Request, res: Response) => {
-        let data: Job;
+        let data: Job;        
         let data2: Responsibility[];
+        let isLoggedIn: boolean = req.session.isLoggedIn;
```
Comment (eoinmcbennett) [inline controller/jobController.ts]: change let to const here

File: controller/jobController.ts [hunk 14]
```diff
@@ -1,37 +1,61 @@
...
 
         try {
             data = await jobService.getJobSpecById(req.params.id);
             data2 = await jobService.getRoleResponsibilityById(req.params.id);
+            isLoggedIn = req.session.isLoggedIn;
```
Comment (eoinmcbennett) [inline controller/jobController.ts]: No need to set this again

File: service/authService.ts [hunk 15]
```diff
@@ -0,0 +1,12 @@
+import { Login } from "../model/auth";
+const axios = require("axios");
+
+module.exports.login = async function (login: Login): Promise<void> {
+    try {
+        const response = await axios.post("http://localhost:8080/api/login", login);
```
Comment (eoinmcbennett) [inline service/authService.ts]: Not gonna work for AWS

File: views/index.html [hunk 16]
```diff
@@ -10,53 +10,73 @@
...
             </div>
+            {% else %}
+            <div class="navbar-collapse" id="navbarNav">
+                <ul class="navbar-nav mr-auto">
+                    <li class="nav-item ml-4">
+                    </li>
+                    <li class="nav-item ml-2">
+                    </li>
+                </ul>
+            </div>
```
Comment (eoinmcbennett) [inline views/index.html]: Don't display an empty list, tell the user there is nothing available

File: views/login.html [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+                                    {{ errormessage }}
+                                  </div>
+                                {% endif %}
+                              </div>
+                              <form action="/login" method="post">
+                                <div class="form-outline form-white mb-5">
+                                  <input type="email" id="username" name="username" class="form-control form-control-lg shadow-sm" placeholder="Email" value="{{ username }}" />
+                                </div>
+                                <div class="form-outline form-white mt-5 mb-5">
+                                  <input type="password" id="password" name="password" class="form-control form-control-lg shadow-sm" placeholder="Password" value="{{ password }}" />
```
Comment (eoinmcbennett) [inline views/login.html]: Don't give the user back their password on a failed login attempt

File: views/login.html [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+                              <div class="container">
+                                {% if errormessage %}
+                                  <div id="login-error" class="alert alert-danger" role="alert">
+                                    {{ errormessage }}
+                                  </div>
+                                {% endif %}
+                              </div>
+                              <form action="/login" method="post">
+                                <div class="form-outline form-white mb-5">
+                                  <input type="email" id="username" name="username" class="form-control form-control-lg shadow-sm" placeholder="Email" value="{{ username }}" />
```
Comment (eoinmcbennett) [inline views/login.html]: is it email or username

**Review by eoinmcbennett — CHANGES_REQUESTED** [review]

File: app.ts (hunk 2 above)
Comment (eoinmcbennett) [inline app.ts]: Also your session runs out every minute, the user will be very frustrated if they get logged out ever minute

File: views/login.html (hunk 18 above)
Comment (CharlesM988) [reply views/login.html]: @eoinmcbennett The user ID in the database is 'Username' but the input type is in the email format. The email address is the username. Does that make sense?

Comment (eoinmcbennett) [reply views/login.html]: Yeah that makes sense, although it probably would have been best to call this email consistently. Don't worry about it  :)

Comment (CharlesM988) [reply views/login.html]: > Yeah that makes sense, although it probably would have been best to call this email consistently. Don't worry about it :)

Agreed!

File: controller/jobController.ts (hunk 8 above)
Comment (CharlesM988) [reply controller/jobController.ts]: @eoinmcbennett This is needed because the Index updates to a different state when the user is logged in. 

This gives the user additional options in the nav bar to choose between Admin or Employee. This code is redirecting from the index to 'job-roles'.

Screenshot 2024-01-09 at 14 26 02

File: config.ts [hunk 19]
```diff
@@ -0,0 +1,12 @@
+import { randomBytes } from "crypto";
+
+const generateRandomKey = () => {
+    const randomBytesResult = randomBytes(32);
+    return randomBytesResult.toString("hex");
+};
+
+const config = {
+    APP_SECRET_KEY: generateRandomKey()
```
Comment (eoinmcbennett) [inline config.ts]: You'll still need to check for your environment variable, if that doesn't exist then you need generate a random key.

**Review by eoinmcbennett — CHANGES_REQUESTED** [review]
Looks great, just one small change needed that could break your AWS deployments.

**Review by eoinmcbennett — APPROVED** [review]
Looks good, great work.
